Mathlib/Analysis/SpecialFunctions/Artanh.lean

```diff
@@ -0,0 +1,150 @@
+-/
+
+@[expose] public section
+
+
+noncomputable section
+
+open Function Filter Set
+
+open scoped Topology
+
+namespace Real
+
+variable {x y : ℝ}
+
+/-- `artanh` is defined using a logarithm, `artanh x = log √((1 + x) / (1 - x))`. -/
+@[pp_nodot]
+def artanh (x : ℝ) :=
+  log √((1 + x) / (1 - x))
+
+theorem artanh_eq_half_log {x : ℝ} (hx : x ∈ Icc (-1) 1) :
+    artanh x = 1 / 2 * log ((1 + x) / (1 - x)) := by
+  rw [artanh, log_sqrt <| div_nonneg (by grind) (by grind), one_div_mul_eq_div]
+
+theorem exp_artanh {x : ℝ} (hx : x ∈ Ioo (-1) 1) : exp (artanh x) = √((1 + x) / (1 - x)) :=
+  exp_log <| sqrt_pos_of_pos <| div_pos (by grind) (by grind)
+
+@[simp]
+theorem artanh_zero : artanh 0 = 0 := by simp [artanh]
+
+theorem sinh_artanh {x : ℝ} (hx : x ∈ Ioo (-1) 1) : sinh (artanh x) = x / √(1 - x ^ 2) := by
+  have : 0 < √((1 + x) / (1 - x)) := sqrt_pos_of_pos <| div_pos (by grind) (by grind)
+  have := sq_sub_sq 1 x
+  grind [artanh, sinh_eq, exp_neg, exp_log, sqrt_div, sqrt_mul]
+
+theorem cosh_artanh {x : ℝ} (hx : x ∈ Ioo (-1) 1) : cosh (artanh x) = 1 / √(1 - x ^ 2) := by
+  have : 0 < √((1 + x) / (1 - x)) := sqrt_pos_of_pos <| div_pos (by grind) (by grind)
+  have := sq_sub_sq 1 x
+  grind [artanh, cosh_eq, exp_neg, exp_log, sqrt_div, sqrt_mul]
+
+/-- `artanh` is the right inverse of `tanh` over (-1, 1). -/
+theorem tanh_artanh {x : ℝ} (hx : x ∈ Ioo (-1) 1) : tanh (artanh x) = x := by
+  have := sq_sub_sq 1 x
+  grind [tanh_eq_sinh_div_cosh, sinh_artanh, cosh_artanh, sqrt_ne_zero', mul_pos]
+
+theorem tanh_lt_one (x : ℝ) : tanh x < 1 := by
+  rw [tanh_eq]
+  field_simp
+  grind [exp_pos]
+
+theorem neg_one_lt_tanh (x : ℝ) : -1 < tanh x := by
+  rw [tanh_eq]
+  field_simp
+  grind [exp_pos]
+
+/-- `artanh` is the left inverse of `tanh`. -/
+theorem artanh_tanh (x : ℝ) : artanh (tanh x) = x := by
+  have h : 0 < (1 + tanh x) / (1 - tanh x) :=
+    div_pos (by grind [neg_one_lt_tanh]) (by grind [tanh_lt_one])
+  rw [artanh, ← exp_eq_exp, exp_log (sqrt_pos_of_pos h),
+    ← sq_eq_sq₀ (le_of_lt <| sqrt_pos_of_pos h) (exp_nonneg x),
+    sq_sqrt (le_of_lt h), tanh_eq, exp_neg]
+  field
+
+theorem strictMonoOn_one_add_div_one_sub :
+    StrictMonoOn (fun (x : ℝ) => (1 + x) / (1 - x)) (Ioo (-1) 1) := by
+  intro x hx y hy h
+  field_simp [show 0 < 1 - x by grind, show 0 < 1 - y by grind]
+  grind
+
+theorem strictMonoOn_artanh : StrictMonoOn artanh (Ioo (-1) 1) := by
+  apply strictMonoOn_log.comp ?_ fun x hx ↦ sqrt_pos_of_pos <| div_pos (by grind) (by grind)
+  apply strictMonoOn_sqrt.comp strictMonoOn_one_add_div_one_sub
+    fun x hx ↦ show 0 ≤ (1 + x) / (1 - x) by exact div_nonneg (by grind) (by grind)
+
+theorem artanh_le_artanh {x y : ℝ} (hx : x ∈ Ioo (-1) 1) (hy : y ∈ Ioo (-1) 1) :
+    artanh x ≤ artanh y ↔ x ≤ y :=
+  strictMonoOn_artanh.le_iff_le hx hy
+
+theorem artanh_lt_artanh {x y : ℝ} (hx : x ∈ Ioo (-1) 1) (hy : y ∈ Ioo (-1) 1) :
+    artanh x < artanh y ↔ x < y :=
+  strictMonoOn_artanh.lt_iff_lt hx hy
```

Comment:
I would add one way variants of this that take fewer hypotheses. that is, `(hx : -1 < x) (hy : y < 1) (hxy : x ≤ y) : artanh x ≤ artanh y`, and similarly for `<`, and give the existing names to these new declarations. You can then rename the current ones to `artanh_le_artanh_iff`.